File: my_py_pkg/my_py_pkg/number_counter.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from example_interfaces.msg import Int64
from example_interfaces.srv import SetBool


class NumberCounterNode(Node):
    def __init__(self):
        super().__init__("number_counter") 
        #contador variable global
        self.counter_ = 0
        #iniciar la suscripcion al topic
        self.suscriber_ = self.create_subscription(Int64, "number", self.callback_number_publisher,10)
        #iniciarl el publisher
        self.publisher_ = self.create_publisher(Int64, "number_count", 10) #datatype  , name , qeue size
        #self.timer_ = self.create_timer(0.5, self.publish_number_count) #llama al metodo cada 1/2s seg
        #logs
        self.get_logger().info("Number Counter node has been started")
        #creamos el server
        self.server_ = self.create_service(SetBool, "reset_counter", self.callback_reset_counter )

    def callback_number_publisher(self, msg: Int64):
        self.get_logger().info("Received: " + str(msg.data))
        self.counter_+=1
        #self.get_logger().info("Counter: " + str(self.counter_))
        #al recibir el dato del topic publicar
        new_msg = Int64()
        new_msg.data = self.counter_
        self.publisher_.publish(new_msg)

    def publish_number_count(self):
        #1 crear message
        msg = Int64()
        msg.data = self.counter_
        #2 publicar
        self.publisher_.publish(msg)

    def callback_reset_counter(self, request: SetBool.Request, response: SetBool.Response):
        request_ = request.data
        response.message = "received"
        response.success = False
        if(request_):
            self.counter_ = 0
            self.get_logger().info("received request TRUE")
            response.success = True
            response.message = "received True msg"
        else:
            self.get_logger().info("received request FALSE")
            response.success = True
            response.message = "received False msg"
        return response

 
def main(args=None):
    rclpy.init(args=args)
    node = NumberCounterNode()  
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == "__main__":
    main()

File: my_cpp_pkg/src/number_counter.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"
#include "example_interfaces/srv/set_bool.hpp"


using namespace std::placeholders;
using namespace std::chrono_literals;

class NumberCounterNode : public rclcpp::Node  
{
public:
    NumberCounterNode() : Node("number_counter"), counter_(0)   
    {
        //iniciar la suscripcion
        subscriber_ = this->create_subscription<example_interfaces::msg::Int64>(
            "number",
            10,
            std::bind(&NumberCounterNode::callbackNumberPublisher, this, _1));

        //iniciar publiser
        publisher_ = this -> create_publisher<example_interfaces::msg::Int64>("number_count",10);
        //timer_ = this -> create_wall_timer(0.5s, std::bind(&NumberCounterNode::publishCounter, this));  //publicar con timer

        //iniciar server
        server_ = this-> create_service<example_interfaces::srv::SetBool>(
            "reset_counter"
            ,std::bind(&NumberCounterNode::callbackResetCounter,
            this,
            _1,
            _2
            ));

        //logs        
        RCLCPP_INFO(this->get_logger(), "Number Counter Node has been started 2");
        
    }
private:

    void callbackNumberPublisher(const example_interfaces::msg::Int64::SharedPtr msg)
    {
        RCLCPP_INFO(this->get_logger(), "%ld", msg->data);
        counter_++;
        //publicar cuando se llame a esta metodo
        auto new_msg = example_interfaces::msg::Int64();
        new_msg.data = std::int64_t(counter_);
        publisher_ -> publish(new_msg);

    }

    void publishCounter()
    {
        auto msg = example_interfaces::msg::Int64();
        msg.data = std::int64_t(counter_);
        publisher_ -> publish(msg);
        
    }

    void callbackResetCounter(
        const  example_interfaces::srv::SetBool::Request::SharedPtr request,
        const example_interfaces::srv::SetBool::Response::SharedPtr response
    )
        {            
            if(request->data)
            {
                RCLCPP_INFO(this->get_logger(), "Received True");
                response->message = "Received True";
                response->success = true;
                counter_ = 0;
            }else{
                RCLCPP_INFO(this->get_logger(), "Received False");
                response->message = "Received False";
                response->success = true;
            }
        }


    rclcpp::Subscription<example_interfaces::msg::Int64>::SharedPtr subscriber_;

    rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
    int counter_;

    rclcpp::Service<example_interfaces::srv::SetBool>::SharedPtr server_;



};

int main(int argsc, char **argv){
    rclcpp::init(argsc, argv);
    auto node = std::make_shared<NumberCounterNode>(); 
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}